### PR TITLE
make third party repo image names configurable

### PR DESCRIPTION
**Pull request by anandkumarpatel**
Some image names have been hardcoded in helm templates. 
This PR makes all the image names configurable via helm values. 

I have defaulted them to original values in their respective `values.yaml` to ensure backward compatibility. 

why?
In my use-case, all our images have to come from our image repository and our image repository has strict naming conventions for our images. This change will allow me to change image names without having to fork and change image names in the source.

### CODE REVIEW

**Comment by howardjohn** [comment]
@anandkumarpatel 

We are moving work on istio installer into our "next gen" installer at https://github.com/istio/installer. The new installer will have first class support for [Kustomize](https://github.com/kubernetes-sigs/kustomize), which is perfect for your use case of changing the image name. If you use that you can just create an overlay to override the image name.

However I don't think we should make this change in the Helm templates. I know we have some images with it, but we consider it a mistake that we haven't reverted due to backwards compatibility concerns.

I completely understand that your use case requires changing them, but we can't add an option for everything or it gets too complicated. I would highly recommend you look into using https://github.com/istio/installer which is going to be alpha released with Istio 1.2 (coming very soon)

**Comment by sdake** [comment]
@howardjohn we do have a policy for the 1.2 installer that third party component's names can be customized in the template (in the unlikely event the upstream changes result in an image name change, which we have actually experienced requiring a z stream).  Images that come from docker.io/istio, which we control, however, are hardcoded intentionally.

@anandkumarpatel thank you for your work and the PR.  This isn't the first PR we have seen that does this work, and it probably won't be the last.  As @howardjohn explained in https://github.com/istio/istio/pull/14815#issuecomment-502152032 we plan to address this in the next-gen installer (and if the stars align, deprecate this installer in the 1.3 release).  As a result of the deprecation decision by the technical oversight committee and in agreement with the environments workgroup leads (@sdake and @costinm), we have effectively frozen this directory for master unless a change is required to make the installer function properly.  No new features are permitted.  I do respect your requirements and we are happy to make this type of functionality available in the next generation installer.

Please read https://discuss.istio.io/t/maintainers-please-read-installer-next-plans-for-1-2/2446 for more details.

Cheers
-steve

File: install/kubernetes/helm/istio/charts/certmanager/templates/deployment.yaml [hunk 1]
```diff
@@ -35,7 +35,7 @@ spec:
 {{- end }}
       containers:
       - name: certmanager
-        image: "{{ .Values.hub }}/cert-manager-controller:{{ .Values.tag }}"
+        image: "{{ .Values.hub }}/{{ .Values.image }}:{{ .Values.tag }}"
```
Comment (sdake) [inline install/kubernetes/helm/istio/charts/certmanager/templates/deployment.yaml]: this looks like a third party dependency that was missed when we cleaned up the customizable values.  I know this isn't of much help for your use case, but would accept a PR that addresses this specific third party dependency.

File: install/kubernetes/helm/istio/charts/kiali/templates/deployment.yaml [hunk 2]
```diff
@@ -32,7 +32,7 @@ spec:
       priorityClassName: "{{ .Values.global.priorityClassName }}"
 {{- end }}
       containers:
-      - image: "{{ .Values.hub }}/kiali:{{ .Values.tag }}"
+      - image: "{{ .Values.hub }}/{{ .Values.image }}:{{ .Values.tag }}"
```
Comment (sdake) [inline install/kubernetes/helm/istio/charts/kiali/templates/deployment.yaml]: looks like someone went in and re-hardcoded the third party dependencies...

File: install/kubernetes/helm/istio/charts/prometheus/templates/deployment.yaml [hunk 3]
```diff
@@ -30,7 +30,7 @@ spec:
 {{- end }}
       containers:
         - name: prometheus
-          image: "{{ .Values.hub }}/prometheus:{{ .Values.tag }}"
+          image: "{{ .Values.hub }}/{{ .Values.image }}:{{ .Values.tag }}"
```
Comment (sdake) [inline install/kubernetes/helm/istio/charts/prometheus/templates/deployment.yaml]: same story here.

File: install/kubernetes/helm/istio/charts/tracing/templates/deployment-jaeger.yaml [hunk 4]
```diff
@@ -42,7 +42,7 @@ spec:
 {{- end }}
       containers:
         - name: jaeger
-          image: "{{ .Values.jaeger.hub }}/all-in-one:{{ .Values.jaeger.tag }}"
+          image: "{{ .Values.jaeger.hub }}/{{ .Values.jaeger.image }}:{{ .Values.jaeger.tag }}"
```
Comment (sdake) [inline install/kubernetes/helm/istio/charts/tracing/templates/deployment-jaeger.yaml]: zug same story here...

**Comment by anandkumarpatel** [comment]
@sdake updated 3rd party images. Would you accept a PR for others as you suggested with https://github.com/istio/istio/blob/master/install/kubernetes/helm/istio/charts/pilot/templates/deployment.yaml#L50 ?

File: install/kubernetes/helm/istio/charts/tracing/templates/deployment-jaeger.yaml (hunk 4 above)
Comment (anandkumarpatel) [reply install/kubernetes/helm/istio/charts/tracing/templates/deployment-jaeger.yaml]: done

**Comment by anandkumarpatel** [comment]
Friendly Monday morning poke :) 
@sdake @costinm

**Comment by costinm** [comment]
Few concerns - and one recommendation if you really need this.

The main problem we have with installer (and possibly Istio) is that we have far too many options for anyone to fully understand. And almost none of the options are tested - including this PR. We do require tests for code changes - but apparently install and config which can have a far bigger impact is forgotten.
And the fact we have so many documented options is also used as an excuse ( nobody can test that many options and combinations ). 

I understand you use case - and that of most other option added. And we do want to support them,
however not by adding every single word from a k8s template to values.yaml.

First - please make the changes in the istio/install repository for the new installer. After that if needed we can also backport to old installer ( in case we'll still ship it in 1.3).

Second - consider using the kustomize step of the new installer, if possible. We now support a combination of template and kustomize, by kustomizing after the template rendering. If you use a CI/CD system with 'helm template' - it's just an extra command to apply kustomize.  This has first class support for overriding images and tags, and it's pretty clean.

If you can't use kustomize ( for example use helm tiller ) - it is possible to have customisable fields in the templates that are not exposed in values.yaml. Like with Kustomize - many things can be changed by advanced users without exposing it to typical users. To do that - just use the default value in the helm template. 

The goal is to have a very small 'values.yaml' for most users, with actual tests for the few values we need there - and use kustomize for advanced uses.

**Comment by anandkumarpatel** [comment]
@costinm I understand your concerns. I made the PR for the installer https://github.com/istio/installer/pull/253.

We can't use kustomize right now since we use helm with tiller in our pipeline. 
If I changed this to use `| defaults` instead of putting it in `values.yaml` would it likely be merged here?

**Comment by costinm** [comment]
Yes, I think defaults is one option.

We also discussed having a separate values.yaml with 'defaults' and
corner-case, uncommon settings - and a 'mesh.yaml' with the small set of
settings commonly used.

But given that image is used in istio/istio, I think I'll just approve this
as is. Give me a bit of time to think a bit more...

**Comment by sdake** [comment]
@costinm re this specific PR, this fixes a very specific problem we have with the installation - that is that third-party hub **AND** images can change upstream.  As per policy in several reviews, we have decided (I thought) that third party customization was fine, but customizing istio image names was not fine.  I know its inconsistent, however, we control istio image names, and I gave the green light for this work already based upon these decisions that have been agreed upon in past reviews.

Regarding testing, that isn't a problem easily solveable nor should we dump that on @anandkumarpatel just because the rest of the project has not properly sorted out a testing strategy for our installation...

I can approve this specific PR but would like you to agree.

Cheers
-steve

**Comment by sdake** [comment]
semi-related: https://github.com/istio/istio/pull/15195/files

**Comment by sdake** [comment]
Can you also take care of the grafana third party hub/image?  ATM grafana is blocking caching of upstream images, and we need the hub to be independent of the image name.  It would be nice if this was one pr to backport and not two (if we choose to backport to a z stream).

**Comment by costinm** [comment]
I did think a bit more about this - and the implications on operator and kustomize work we are doing.
This doc in particular, on how this would work for people who rely on kustomizations: 
https://github.com/kubernetes-sigs/kustomize/blob/master/docs/fields.md#images
Also keeping in mind that values.yaml is getting transformed into a properly schemed configuration, and will be used by operator as 'mesh config'.

How about adopting the same style with kustomize: we allow 'tag' and 'name' to be customized, with 'name' defined as the combination of hub and image name ? 

Ideally we would add an "images" section, following the same syntax as kustomize - but I don't know if we can do this in helm.
For people using Kustomize, they would use something like:
```
images:
- name: istio-pilot
  newName: my-registry/my-pilot
  newTag: v1
```

Can helm template use this as an input ? 

On testing: I am not suggesting full end-to-end tests for this, but some basic "generate the config with a specific setting with helm template, grep the output to verify it has the expected result". I believe there are few other ways to 'unit test' helm templates ( @Stono mentioned one he is using ). This can also serve as an example, like any good test. 

In particular I am looking for confirmation that when a feature like this is added - image selection - we can review the usability and how it looks in both helm and kustomize - and for that we need at least some examples included in the PR. And a simple test is also an example.

**Comment by costinm** [comment]
To clarify: I would expect configuring the image name with Kustomize will use the kustomize style, and the test for changing the images will kustomize will use a Kustomize file. 
Given that we plan the 'easy' installation to be based on kustomize and 'kubectl -k ..' (and operator in future) we can't ignore this.

**Comment by sdake** [comment]
> I did think a bit more about this - and the implications on operator and kustomize work we are doing.
> This doc in particular, on how this would work for people who rely on kustomizations:
> https://github.com/kubernetes-sigs/kustomize/blob/master/docs/fields.md#images
> Also keeping in mind that values.yaml is getting transformed into a properly schemed configuration, and will be used by operator as 'mesh config'.
> 
> How about adopting the same style with kustomize: we allow 'tag' and 'name' to be customized, with 'name' defined as the combination of hub and image name ?
> 
> Ideally we would add an "images" section, following the same syntax as kustomize - but I don't know if we can do this in helm.
> For people using Kustomize, they would use something like:
> 
> ```
> images:
> - name: istio-pilot
>   newName: my-registry/my-pilot
>   newTag: v1
> ```
> 
> Can helm template use this as an input ?

I think it may be possible, although @morvencao may have more insight into how to do the transformation.  If it were in the subchart values.yaml, it would work straight away.  Not sure if we can import kustomize blobs directly into values files, but it may be possible with a tool (that inserts global.).

Any simpler ways @morvencao ?

**Comment by costinm** [comment]
BTW - my comment applies primarily on the istio/installer use of this feature - this PR in particular is consistent with what we are doing for the other images, so no problem with it ( and since we don't test changing the pilot image - I'm not going to ask for it ). 

/lgtm
/approve 
(but you have a conflict )

**Comment by sdake** [comment]
thanks @costinm!  Also when rectifying the merge conflict, can you address grafana in this same fashion as the rest of this PR?

Cheers
-steve

**Comment by costinm** [comment]
We should move the discussion about this to a feature request in installer... 
The 'feature' is to have a supported, documented and tested way to customize images and tags, for all 3 installation flows ( kustomize, helm tiller, operator)

**Comment by anandkumarpatel** [comment]
@costinm updated with master again. LMK if there if you want anything else in this PR

**Comment by anandkumarpatel** [comment]
Also @sdake grafana chart abstracts this out in a different way. 
```yaml
grafana:
  image:
    repository: grafana/grafana
    tag: 6.1.6
``` 
Would you like me to split the `grafana.image.repository` key into `grafana.image.hub` and `grafana.image.image` to match the others? (this can cause some backward-incompatible issue if people are setting this via cli with `--set`

**Pull request merged**